### PR TITLE
Specify macaroon root key during wallet init

### PR DESCRIPTION
**Pull request by sangaman**
## Change Description

This set of commits adds the ability to specify the macaroon root key to be used when initializing a wallet. This allows for deterministically generating macaroons for a new node/wallet. In tandem with specifying the wallet seed, a wallet can be initialized without requiring secrets to be randomly generated by lnd upon wallet initialization.

## Steps to Test

These are the steps I used to manually test this change.

1. Start lnd from a fresh start with no wallet.
2. Use the commands below to initialize the wallet with a root macaroon key we randomly generate outside of lnd.

```bash
CA_CERT=~/.lnd/tls.cert

WALLETPASS=somepass
WALLETPASS_BASE64=$(printf $WALLETPASS|base64|tr -d '\n\r')

ROOTKEY_BASE64=$(dd status=none if=/dev/urandom bs=32 count=1 | base64)
echo $ROOTKEY_BASE64

CIPHER_SEED=$(curl -s --cacert "$CA_CERT" -X GET https://localhost:8080/v1/genseed | jq -c -r '.cipher_seed_mnemonic')

INITWALLET_REQ='{"wallet_password":"'$WALLETPASS_BASE64'", "macaroon_root_key":"'$ROOTKEY_BASE64'", "cipher_seed_mnemonic":'$CIPHER_SEED'}'

curl -s --cacert "$CA_CERT" -X POST -d "$INITWALLET_REQ" https://localhost:8080/v1/initwallet > /dev/null
```
3. Verify that you can use the newly generated `admin.macaroon` to make `lncli` commands to lnd.
4. Rename `admin.macaroon` to `admin.macaroon.old` and move it outside of the lnd dir.
5. Delete the lnd `data` dir
6. Start lnd again from another fresh start.
7. Run the same commands as in step 2 to init the wallet but assign `ROOTKEY_BASE64` to the value that was `echo`ed in step 2, e.g. `ROOTKEY_BASE64="3BVV9mZmqAv0/BUS2jOvKLNDGhxssLTAfVlUYGeSAUw="`
8. Verify that you can make `lncli` commands with the `admin.macaroon.old` file you saved from step 4.

### CODE REVIEW

File: config_builder.go
Comment (guggero) [inline config_builder.go:430]: This return should probably be wrapped by a `if err != nil {` condition?

File: macaroons/store.go
Comment (guggero) [inline macaroons/store.go:351]: Encrypt root key outside of DB transaction to block it for a shorter amount of time?

**Review by guggero — COMMENTED** [review]
Did a quick first pass and haven't tested the functionality yet. But this looks pretty good and quite useful too.

File: config_builder.go
Comment (sangaman) [reply config_builder.go:430]: Thanks for the catch, this was a lint error I addressed right before committing (previously I was ignoring the return value of `SetRootKey` and I added the `err :=` and `return` but forgot the `if`.

File: walletunlocker/service.go
Comment (guggero) [inline walletunlocker/service.go:411]: nit: somehow it's easier to read `len() > 0 && len() != 32`. But might just be me. We should replace the magic number by `macaroons.RootKeyLen` though.

File: macaroons/store.go
Comment (guggero) [inline macaroons/store.go:341]: Same here re magic number.

File: macaroons/store_test.go
Comment (guggero) [inline macaroons/store_test.go:187]: We don't actually read the current key. Maybe we should, just to make sure it's not yet set to the one we're about to set?

File: config_builder.go
Comment (guggero) [inline config_builder.go:428]: nit: line slightly too long when using 8 spaces for the tab character.

**Review by guggero — APPROVED** [review]
tACK, very nice feature for test environments!

Could be accompanied by a simple integration test, but not blocking IMO.

**Review by ellemouton — APPROVED** [review]
LGTM pending a rebase and fixing the nits mentioned by @guggero 👍

File: walletunlocker/service.go
Comment (sangaman) [reply walletunlocker/service.go:411]: Looking around the codebase, I see roughly equal usage of `len() > 0 && len() != 32` and `len() != 0 && len() != 32` and I believe I copied this from one such case of the latter, but I'll change it here since it doesn't matter to me.

File: macaroons/store_test.go
Comment (sangaman) [reply macaroons/store_test.go:187]: I just removed this comment for now, personally I don't see why lnd should prevent setting the root key to its current value. But if you or someone feels strongly about not allowing that I can add that logic in.

File: walletunlocker/service.go
Comment (guggero) [inline walletunlocker/service.go:415]: Missing `len` function call. And also line too long.

File: config_builder.go
Comment (guggero) [inline config_builder.go:426]: nit: missing full stop here and in a few other comments.

File: config_builder.go
Comment (guggero) [inline config_builder.go:429]: nit: closing parenthesis should be on new line.

**Review by guggero — COMMENTED** [review]
Thanks for the update. Only nits left.

@Roasbeef do you think we can get this in for 0.15? Or should we move it to 0.15.1?

**Comment by sangaman** [comment]
Fixed latest round of feedback.

File: walletunlocker/service.go
Comment (guggero) [inline walletunlocker/service.go:414]: nit: missing a space in the error message after "exactly".

**Review by guggero — APPROVED** [review]
LGTM, thanks a lot for the quick fixes! I'm hopeful this can get in with 0.15 as we're still in the RC process.

**Comment by sangaman** [comment]
Hi @guggero - is there anything I can/should do  before this gets merged? I just don't want this to get forgotten or miss any merge windows.

**Comment by sangaman** [comment]
I actually just went ahead and moved the release notes from 0.15.0 to 0.15.1 - hoping it can make it in that!

**Comment by sangaman** [comment]
OK I actually had to make some changes to this PR (specifically the `macaroons: add SetRootKey method` commit) after rebasing on top of #6509. Since that's flagged for the 0.16 release, and since this PR has the 0.16 release tag as well, I moved the release change notes into 0.16 here as well. 

I'm looking forward to getting this into a release tho since we're currently using a patched version of lnd with the commits from this PR since we need the macaroon root key functionality. Thanks!

File: macaroons/service.go
Comment (guggero) [inline macaroons/service.go:314]: I'm not 100% sure it's a good idea to return `nil` here, since that means we didn't do anything. But looking at all the other methods here, that seems to be the pattern. So not going to block the PR on that, perhaps we should take another look at this whole service in another PR.

**Comment by guggero** [comment]
Thanks for the reminder! We would've come across the PR again when focusing on the 0.16 board. But better to get things that are ready in early.

File: macaroons/service.go
Comment (sangaman) [reply macaroons/service.go:314]: Good point. Indeed I just went with the pattern that other methods were using.

**Pull request merged**
